### ISSUE TITLE
Using parallel in a laravel app

### ISSUE THREAD

**Issue by Onepamopa** [krakjoe/parallel#291]
Hey there,

I'm trying to run a few "tasks" in parallel (each of the tasks takes between 2~5 seconds which totals to about 16-17 seconds before the function returns (and the page is rendered)).

This is what I've tried, however, the runtime doesn't know what Collection and Carbon classes are, so it spits errors:
local.ERROR: parallel\Runtime\Type\Unavailable::App\Http\Controllers\Storefront\{closure}(): Argument #1 ($allProducts) must be of type Illuminate\Support\Collection, parallel\Runtime\Type\Unavailable given
&
illegal parameter (Carbon\Carbon) passed to task at argument 2

What am I doing wrong?? I've tried passing the bootstrap\autoload as bootstrap for the Runtime(), but "bootstrapping failed".

```php
$tasks = [
    'conditionCounts' => function (\Illuminate\Support\Collection $allProducts) {
        return array_count_values($allProducts->pluck('condition')->toArray());
    },
    'hasOffers' => function (\Illuminate\Support\Collection $allProducts, \Carbon\Carbon $now) {
        return $allProducts->filter(function ($product) use ($now) {
            return $product['offer_price'] > 0 && $product['offer_start'] < $now && $product['offer_end'] > $now;
        })->count();
    },
    'hasFreeShipping' => function (\Illuminate\Support\Collection $allProducts) {
        return $allProducts->where('free_shipping', 1)->count();
    },
    'newArrivals' => function (\Illuminate\Support\Collection $allProducts, \Carbon\Carbon $now) {
        return $allProducts->where('created_at', '>', $now->subDays(config('system.filter.new_arrival', 7)))->count();
    },
];

$parallel = new Runtime();
$promises = [];
$now = \Carbon\Carbon::now();
foreach ($tasks as $key => $operation) {
    $promises[$key] = $parallel->run($operation, [$all_products, $now]);
}

$results = [];
foreach ($promises as $key => $promise) {
    $results[$key] = $promise->value();
}
```

**Comment by goetas** [comment]
After digging a little, your code should be:

```php
$promises = [];
$parallels = [];

$now = \Carbon\Carbon::now();
foreach ($tasks as $key => $operation) {
    $parallels[] = $parallel = new Runtime();
    $promises[$key] = $parallel->run($operation, [$all_products, $now]);
}

foreach ($parallels as $parallel) {
    $parallel->close();
}

$results = [];
foreach ($promises as $key => $promise) {
    $results[$key] = $promise->value();
}
```

**Comment by goetas** [comment]
However, if `$all_products` is a database backed resource, you need to make sure that a separtate connection gets created for each thread

**Comment by Onepamopa** [comment]
Still getting the same error:
 illegal parameter (Carbon\Carbon) passed to task at argument 2 {"exception":"[object] (parallel\\Runtime\\Error\\IllegalParameter(code: 0): illegal parameter (Carbon\\Carbon) passed to task at argument 2 at

And if I remove $now (and the tasks that require it) - 
local.ERROR: parallel\Runtime\Type\Unavailable::App\Http\Controllers\Storefront{closure}(): Argument https://github.com/krakjoe/parallel/issues/1 ($allProducts) must be of type Illuminate\Support\Collection, parallel\Runtime\Type\Unavailable given

**Comment by goetas** [comment]
See https://github.com/krakjoe/parallel/issues/73

**Comment by goetas** [comment]
more specifically https://github.com/krakjoe/parallel/issues/73#issuecomment-504656960

**Comment by goetas** [comment]
Carbon extends DateTime, and it seems not allowed

**Comment by Onepamopa** [comment]
So, in essence what I need can't be done using parallel

**Comment by goetas** [comment]
you could, but you need to build `$all_products` inside the closure instead of passing it as argument

**Comment by Onepamopa** [comment]
The idea of passing it is because it's a cached result form a SQL query which by itself takes 1.33 minutes to execute (hence the caching). The result is a Laravel collection (which also doesn't get recognized/isn't allowed).

When retrieving it from cache it takes ~2-3 seconds:
if (Cache::has('large_data')) {
  $all_products = Cache::get('large_data');
} else {
  $all_products = Cache::remember('large_data', 86400, . . . . . 
}

That's why I have to pass it...

**Comment by michael-rubel** [comment]
The bootstrapping doesn't work with the objective `Runtime` instance like this. You should either explicitly pass the autoload to the `Runtime` object or use functional API: https://www.php.net/manual/en/parallel.run.php

So it becomes:

`artisan`
```php
...

$app = require_once __DIR__.'/bootstrap/app.php';

if (extension_loaded('parallel')) {
    parallel\bootstrap(base_path('bootstrap/parallel.php'));
}

...
```

`bootstrap/parallel.php`
```php
<?php

declare(strict_types=1);

require __DIR__ . '/../vendor/autoload.php';

$app = require_once __DIR__ . '/app.php';

$kernel = $app->make(Illuminate\Contracts\Console\Kernel::class);

$kernel->bootstrap();
```

`app/Commands/YourCommand.php`
```php
parallel\run(function (Collection $records) {
    $service = app(YourService::class);

    $service->insert($records);
}, [$records]);
```

**Comment by hschimpf** [comment]
Hi! I'll like to join this conversation.

I've used `parallel` with Laravel to sync a large database between two systems, and yes, dates are a problem.

One way that I did is to pass the date as a string, then inside the thread _(runtime)_ you re-create the Carbon instance:

```diff
$tasks = [
    // ...
-   'hasOffers' => function (\Illuminate\Support\Collection $allProducts, \Carbon\Carbon $now) {
+   'hasOffers' => function (\Illuminate\Support\Collection $allProducts, string $now) {
+       $now = Carbon::parse($now);
        return $allProducts->filter(function ($product) use ($now) {
            return $product['offer_price'] > 0 && $product['offer_start'] < $now && $product['offer_end'] > $now;
        })->count();
    },
    // ...
];
// ...
$now = \Carbon\Carbon::now();
foreach ($tasks as $key => $operation) {
-   $promises[$key] = $parallel->run($operation, [$all_products, $now]);
+   $promises[$key] = $parallel->run($operation, [$all_products, $now->toString()]);
}
```

---

But, seen your initial code, maybe _(it seems like)_ you don't need parallel at all. Maybe you could convert those calculations into database queries? Since you are only getting a count for all of those.

Indexes are a good friend here. I've working with large datasets of 80k+ products and adding indexes helped me a lot. I went from queries running between ~10-15s down to ~800ms or less.

I don't know about the first one, but it could be that the other three can be converted into something like this?

```php
$hasOffers = Product::where('offer_price', '>', 0)
    ->where('offer_start', '<', $now)
    ->where('offer_end', '>', $end)
    ->count();

$hasFreeShipping = Product::where('free_shipping', 1)->count();

$newArrivals = Product::where('created_at', '>', $now->subDays(config('system.filter.new_arrival', 7)))->count();
```

Adding the right indexes will help the database engine to run those queries fast.

**Comment by krakjoe** [comment]
This doesn't look like a bug to me, it seems to be caused by confusion around bootstrapping.

There's nothing much more to say here, except that the constructor to `Runtime` and functional API equivalent `\parallel\bootstrap` should both behave the same way with the same bootstrap file. If you can find a difference, with a simple, self contained, reproducing script, please open a new issue regarding that (that seems highly unlikely).

Whenever you see `Unavailable` type in exceptions it's because bootstrapping was not sufficient for the runtime, ie, a symbol is unavailable, and the name of the symbol precedes `Unavailable` in the trace.

I'm glad is sparked conversation, but I'm closing the issue now as I don't think there's anything for us to work on here.
